feat(Convolution2DPass): Add 2D convolution kernels as render pass

casra-developers opened the pull request:
The class vtkCustomPostProcessingPass was added. The class vtkCustomPostProcessingPass can be
integrated in into the vtk.js render-pipeline to enable 2D image enhancements on the rendered
images. This works by setting the vtkRenderPass as a delegate pass and then using the
vtkCustomPostProcessingPass as the render pass on the vtkRenderWindow. Any valid
2D-convolution-kernel can be used.
As discussed via email we would be glad if @finetjul could review our changes.
![commit](https://user-images.githubusercontent.com/68997390/88922899-1db8a900-d271-11ea-8933-dc9192168f69.png)

finetjul commented:
@boucaud WDYT ?

finetjul commented:
@casra-developers, this looks great, thanks ! Can you please create a [test](https://github.com/Kitware/vtk-js/tree/master/Sources/Rendering/OpenGL/StickMapper/test) or an [example](https://github.com/Kitware/vtk-js/tree/master/Sources/Filters/General/WindowedSincPolyDataFilter/example)

@boucaud will review the code itself.

boucaud commented on Sources/Rendering/OpenGL/CustomPostProcessingPass/index.js:
```diff
@@ -0,0 +1,317 @@
+const { vtkErrorMacro } = macro;
+// ----------------------------------------------------------------------------
+
+function vtkCustomPostProcessingPass(publicAPI, model) {
```
`vtkCustomKernelPass` or `vtkConvolutionPass` would be more indicative names for what the class does.

casra-developers replied:
We agree that the chosen name is not very descriptive. What about vtk2dConvolutionPass to make it clear that it uses a 2D kernel on a 2D image?

boucaud replied:
I agree, `vtk2DConvolutionPass` sounds good.

boucaud commented on Sources/Rendering/OpenGL/CustomPostProcessingPass/index.js:
```diff
@@ -0,0 +1,317 @@
+      val.traverse(viewNode, publicAPI);
+    });
+
+    // now draw the distorted values
```
`distorted` -> `convolved`

casra-developers replied:
Makes sense we will change this.

boucaud commented on Sources/Rendering/OpenGL/CustomPostProcessingPass/index.js:
```diff
@@ -0,0 +1,317 @@
+      model.convolutionShader !== null &&
+      model.oldKernelDimension !== model.kernelDimension
+    ) {
+      // model.convolutionShader.cleanup(); // todo: what does this do?
```
I think you can uncomment this (it just calls `gl.deleteShader`)

casra-developers replied:
We used to have this active but it caused crashes when changing the kernel size. Do you think we should try to re-enable it or just remove the line entirely?

boucaud replied:
I think it would be interesting to see if 
```
  model.convolutionShader.release();
  viewNode.getShaderCache().setLastShaderBound(null);
  model.convolutionShader.cleanup();
```
Fixes the crash issue.  

If that doesn't work, then you can drop the line.

casra-developers replied:
We tried it with your suggestion but it still crashes. In the most recent commit we omitted the line.

![image](https://user-images.githubusercontent.com/68997390/89035052-8a4baa80-d33a-11ea-9010-3c769083dfce.png)

boucaud replied:
It's ok to just drop that line.

boucaud commented on Sources/Rendering/OpenGL/CustomPostProcessingPass/index.js:
```diff
@@ -0,0 +1,317 @@
+    // finish code
+    shaderCode += [
+      ';',
+      '    gl_FragData[0] = vec4((colorSum / u_kernelWeight).rgb, 1);',
```
Why not keep the alpha value from `u_image` ?

casra-developers replied:
We will try it.

boucaud commented on Sources/Rendering/OpenGL/CustomPostProcessingPass/index.js:
```diff
@@ -0,0 +1,317 @@
+  tris: null,
+  kernel: [0, 0, 0, 0, 1, 0, 0, 0, 0],
+  oldKernelDimension: 3,
+  kernelDimension: 3,
```
`kernelSize`

casra-developers replied:
Strictly speaking kernelSize would be e.g. 3x3. Therefore, we would rather keep the kernelDimension since it indicates one side of the square matrix. Do you have a better idea to make it more intuitive?

boucaud replied:
I think you're right, dimension seems fine.

boucaud commented:
Just a comment: I noticed this duplicates a bit of functionality from `vtkRadialDistortionPass`, which is completely fine.

I do think in the future if more classes like this need to be added, it would make sense to factorize some functionality, such as a `vtkOpenGLRenderPass` that handles the framebuffer and context, and some kind of `vtkOpenGLPostProcessingPass` that handles rendering to a quad, and just provides some overridable methods to build shaders, set uniforms and load other buffers.

casra-developers commented:
@finetjul can you point us towards a tutorial for running and creating examples? We are unable to get the existing ones to work.

boucaud commented:
> @finetjul can you point us towards a tutorial for running and creating examples? We are unable to get the existing ones to work.

[this](https://kitware.github.io/vtk-js/docs/develop_example.html) is probably what you're looking for.

casra-developers commented:
> Just a comment: I noticed this duplicates a bit of functionality from `vtkRadialDistortionPass`, which is completely fine.
> 
> I do think in the future if more classes like this need to be added, it would make sense to factorize some functionality, such as a `vtkOpenGLRenderPass` that handles the framebuffer and context, and some kind of `vtkOpenGLPostProcessingPass` that handles rendering to a quad, and just provides some overridable methods to build shaders, set uniforms and load other buffers.

This is true. We started by investigating how web GL works and then looked for a class that did also work with the framebuffer as a render target of the delegate pass. Since we have very little in terms of vtk.js domain knowledge there is certainly potential to remove redundancy.

boucaud commented on Sources/Rendering/OpenGL/Convolution2dPass/index.js:
```diff
@@ -10,9 +10,9 @@ import { Representation } from 'vtk.js/Sources/Rendering/Core/Property/Constants
+function vtkConvolution2dPass(publicAPI, model) {
   // Set our className
-  model.classHierarchy.push('vtkCustomPostProcessingPass');
+  model.classHierarchy.push('vtkConvolution2dPass');
```
A capital D is more consistent here (`vtkConvolution2DPass`)

casra-developers replied:
Will be changed.

boucaud commented on Sources/Rendering/OpenGL/Convolution2dPass/index.js:
```diff
@@ -216,7 +215,7 @@ function vtkCustomPostProcessingPass(publicAPI, model) {
     shaderCode += [
       ';',
-      '    gl_FragData[0] = vec4((colorSum / u_kernelWeight).rgb, 1);',
+      '    gl_FragData[0] = vec4((colorSum / u_kernelWeight).argb, 1);',
```
I was more thinking of something like `gl_FragData[0] = vec4((colorSum / u_kernelWeight).rgb, texture2D(u_image, tcoord).a);`

Convolving the alpha is probably not desired in most cases.

casra-developers replied:
Good point! We will change it to your suggestion.

casra-developers commented:
> > @finetjul can you point us towards a tutorial for running and creating examples? We are unable to get the existing ones to work.
> 
> [this](https://kitware.github.io/vtk-js/docs/develop_example.html) is probably what you're looking for.

Thank this looks great to start.

casra-developers commented:
> > > @finetjul can you point us towards a tutorial for running and creating examples? We are unable to get the existing ones to work.
> > 
> > 
> > [this](https://kitware.github.io/vtk-js/docs/develop_example.html) is probably what you're looking for.
> 
> Thank this looks great to start.

We have some problems in creating the example. Is there a template for a proper example using vtkForwardPass? Using the code of either ZipHttpReader or VolumeMapperBlendModes (since both show volume data that could be used to illustrate the post-processing) results in compliation errors even if the code remains the same. In our own project we just include the vtk.js file and it runs fine. Our problem is also that myself and the other developer working on this task are not experienced javascript developers but rather have backgrounds in C++, C# and Python. Therefore, we do not really understand how all the build dependencies work.

ERROR in ./Rendering/Convolution2DPass/index.js
Module Error (from ../node_modules/eslint-loader/dist/cjs.js):

C:\Projects\vtk-js\Examples\Rendering\Convolution2DPass\index.js
    1:1   error  Expected exception block, space or tab after '//' in comment  spaced-comment
    1:35  error  Delete `␍`                                                    prettier/prettier
    2:1   error  Delete `␍`                                                    prettier/prettier
    3:1   error  Expected space or tab before '*/' in comment                  spaced-comment
    3:1   error  Expected exception block, space or tab after '/*' in comment  spaced-comment
    3:94  error  Delete `␍`                                                    prettier/prettier
    4:94  error  Delete `␍`                                                    prettier/prettier
    5:77  error  Delete `␍`                                                    prettier/prettier
    6:86  error  Delete `␍`                                                    prettier/prettier
    7:62  error  Delete `␍`                                                    prettier/prettier
    8:76  error  Delete `␍`                                                    prettier/prettier
    9:74  error  Delete `␍`                                                    prettier/prettier
   10:75  error  Delete `␍`                                                    prettier/prettier
   11:1   error  Expected exception block, space or tab after '//' in comment  spaced-comment
   11:87  error  Delete `␍`                                                    prettier/prettier
   12:1   error  Delete `␍`                                                    prettier/prettier
   13:33  error  Delete `␍`                                                    prettier/prettier
   14:1   error  Delete `␍`                                                    prettier/prettier
   15:94  error  Delete `␍`                                                    prettier/prettier
   16:77  error  Delete `␍`                                                    prettier/prettier
   17:86  error  Delete `␍`                                                    prettier/prettier
   18:92  error  Delete `␍`                                                    prettier/prettier
   19:62  error  Delete `␍`                                                    prettier/prettier
   20:74  error  Delete `␍`                                                    prettier/prettier
   21:26  error  Unable to resolve path to module './controller.html'          import/no-unresolved
   21:46  error  Delete `␍`                                                    prettier/prettier
   22:1   error  Delete `␍`                                                    prettier/prettier
   23:80  error  Delete `␍`                                                    prettier/prettier
   24:33  error  Delete `␍`                                                    prettier/prettier
   25:80  error  Delete `␍`                                                    prettier/prettier
   26:1   error  Delete `␍`                                                    prettier/prettier
   27:67  error  Delete `␍`                                                    prettier/prettier
   28:31  error  Delete `␍`                                                    prettier/prettier
   29:4   error  Delete `␍`                                                    prettier/prettier
   30:51  error  Delete `␍`                                                    prettier/prettier
   31:59  error  Delete `␍`                                                    prettier/prettier
   32:1   error  Delete `␍`                                                    prettier/prettier
   33:48  error  Delete `␍`                                                    prettier/prettier
   34:1   error  Delete `␍`                                                    prettier/prettier
   35:80  error  Delete `␍`                                                    prettier/prettier
   36:16  error  Delete `␍`                                                    prettier/prettier
   37:80  error  Delete `␍`                                                    prettier/prettier
   38:62  error  Delete `␍`                                                    prettier/prettier
   39:62  error  Delete `␍`                                                    prettier/prettier
   40:80  error  Delete `␍`                                                    prettier/prettier
   41:1   error  Delete `␍`                                                    prettier/prettier
   42:70  error  Delete `␍`                                                    prettier/prettier
   43:1   error  Delete `␍`                                                    prettier/prettier
   44:39  error  Delete `␍`                                                    prettier/prettier
   45:46  error  Delete `␍`                                                    prettier/prettier
   46:31  error  Delete `␍`                                                    prettier/prettier
   47:25  error  Delete `␍`                                                    prettier/prettier
   48:1   error  Delete `␍`                                                    prettier/prettier
   49:47  error  Delete `␍`                                                    prettier/prettier
   50:49  error  Delete `␍`                                                    prettier/prettier
   51:27  error  Delete `␍`                                                    prettier/prettier
   52:29  error  Delete `␍`                                                    prettier/prettier
   53:25  error  Delete `␍`                                                    prettier/prettier
   54:26  error  Delete `␍`                                                    prettier/prettier
   55:1   error  Delete `␍`                                                    prettier/prettier
   56:54  error  Delete `␍`                                                    prettier/prettier
   57:35  error  Delete `␍`                                                    prettier/prettier
   58:37  error  Delete `␍`                                                    prettier/prettier
   59:33  error  Delete `␍`                                                    prettier/prettier
   60:34  error  Delete `␍`                                                    prettier/prettier
   61:1   error  Delete `␍`                                                    prettier/prettier
   62:54  error  Delete `␍`                                                    prettier/prettier
   63:47  error  Delete `␍`                                                    prettier/prettier
   64:58  error  Delete `␍`                                                    prettier/prettier
   65:52  error  Delete `␍`                                                    prettier/prettier
   66:36  error  Delete `␍`                                                    prettier/prettier
   67:37  error  Delete `␍`                                                    prettier/prettier
   68:37  error  Delete `␍`                                                    prettier/prettier
   69:38  error  Delete `␍`                                                    prettier/prettier
   70:44  error  Delete `␍`                                                    prettier/prettier
   71:1   error  Delete `␍`                                                    prettier/prettier
   72:51  error  Delete `␍`                                                    prettier/prettier
   73:1   error  Delete `␍`                                                    prettier/prettier
   74:34  error  Delete `␍`                                                    prettier/prettier
   75:54  error  Delete `␍`                                                    prettier/prettier
   76:1   error  Delete `␍`                                                    prettier/prettier
   77:34  error  Delete `␍`                                                    prettier/prettier
   78:1   error  Delete `␍`                                                    prettier/prettier
   79:25  error  Delete `␍`                                                    prettier/prettier
   80:2   error  Delete `␍`                                                    prettier/prettier
   81:70  error  Delete `␍`                                                    prettier/prettier
   82:33  error  Delete `␍`                                                    prettier/prettier
   83:31  error  Delete `␍`                                                    prettier/prettier
   84:53  error  Delete `␍`                                                    prettier/prettier
   85:43  error  Delete `␍`                                                    prettier/prettier
   86:28  error  Delete `␍`                                                    prettier/prettier
   87:47  error  Delete `␍`                                                    prettier/prettier
   88:27  error  Delete `␍`                                                    prettier/prettier
   89:1   error  Delete `␍`                                                    prettier/prettier
   90:53  error  Delete `␍`                                                    prettier/prettier
   91:52  error  Delete `␍`                                                    prettier/prettier
   92:6   error  Delete `␍`                                                    prettier/prettier
   93:4   error  Delete `␍`                                                    prettier/prettier
   94:1   error  Delete `␍`                                                    prettier/prettier
   95:63  error  Delete `␍`                                                    prettier/prettier
   96:58  error  Delete `␍`                                                    prettier/prettier
   97:55  error  Delete `␍`                                                    prettier/prettier
   98:63  error  Delete `␍`                                                    prettier/prettier
   99:1   error  Delete `␍`                                                    prettier/prettier
  100:24  error  Delete `␍`                                                    prettier/prettier
  101:24  error  Delete `␍`                                                    prettier/prettier
  102:22  error  Delete `␍`                                                    prettier/prettier
  103:20  error  Delete `␍`                                                    prettier/prettier
  104:28  error  Delete `␍`                                                    prettier/prettier
  105:36  error  Delete `␍`                                                    prettier/prettier

✖ 110 problems (110 errors, 0 warnings)
  109 errors and 0 warnings potentially fixable with the `--fix` option.


ERROR in ./Rendering/Convolution2DPass/index.js
Module not found: Error: Can't resolve './controller.html' in 'C:\Projects\vtk-js\Examples\Rendering\Convolution2DPass'
 @ ./Rendering/Convolution2DPass/index.js 19:0-45 28:33-45

boucaud commented:
For the `VolumeMapperBlendModes` example, you also need to copy/modify the `controller.html` file from the `VolumeMapperBlendModes` directory, (or remove the import in the example).

`npm run reformat` will probably with those prettier errors

casra-developers commented:
> For the `VolumeMapperBlendModes` example, you also need to copy/modify the `controller.html` file from the `VolumeMapperBlendModes` directory, (or remove the import in the example).
> 
> `npm run reformat` will probably help with those prettier errors

Thank we will try it again and see what comes out.

casra-developers commented:
Thanks to the hints from @boucaud we were able to add a working example that allows you to chain together multiple convolution passes and tweak them. Of course adding several edge enhancements on top of each other will end up looking strange, so one possibility would be to choose radio buttons instead of checkboxes and leave the checkbox on the Gaussian blur since this is a common step before edge enhancement operations . Looking forward to the feedback.

casra-developers commented:
Does anybody have an idea why the checks are now failing? I tried to make the naming consistent by switching everywhere from lowercase "d" to uppercase "D". Apparently this was a mistake.

finetjul commented on Sources/Rendering/OpenGL/Convolution2DPass/index.js:
```diff
@@ -0,0 +1,311 @@
+
+  // build vertices etc
+  publicAPI.buildVertexBuffer = () => {
+    // 4 coner points in clipping space in order (x, y, z) where z is always set to -1
```
coner -> corner

finetjul commented on Sources/Rendering/OpenGL/Convolution2DPass/index.js:
```diff
@@ -0,0 +1,311 @@
+      -1,
+      1,
+      1,
+      -1,
```
you could have used 4 lines for readability

floryst replied:
To follow-up: if `npm run reformat` is resulting in values on each line, then put `// prettier-ignore` above line 226.

floryst commented:
CI issues were transient issues. Re-ran CI and now instances report green.

casra-developers commented:
I implemented the changes as requested for better readability and also got rid of the typo in one comment. In the example the edge enhancement filters are now chosen one at a time and not chained together anymore. Thank you @floryst for fixing the CI and also thanks to @finetjul for the additional comments. Many thanks to @boucaud for all the support during this pull-request. I am not sure how to squash the commits but I renamed the feature.

finetjul commented:
This is great, thanks.

I just rebased master into the branch and now waiting for tests to be ran before merging. For information, squashing could be done with `git rebase -i HEAD~11`

jourdain commented:
:tada: This PR is included in version 14.14.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/vtk.js/v/14.14.0)
- [GitHub release](https://github.com/Kitware/vtk-js/releases/tag/v14.14.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
